### PR TITLE
libgpiod & orangepipc support

### PR DESCRIPTION
**Pull request by ladyada**
did up libgpiod-based gpio - turns out there's no pullup support so this isn't quite ready to replace rpi.gpio or bb.gpio since pullups are pretty sweet 

still, got everything else working - all gpio, and uart, i2c and spi. removed some old debug statements too.

### CODE REVIEW

**Comment by ladyada** [comment]
i put down a few reviewers, but its an OR - any of y'all can do it, doesn't have to be all :)

**Comment by brennen** [comment]
> turns out there's no pullup support so this isn't quite ready to replace rpi.gpio or bb.gpio since pullups are pretty sweet

ah, yeah.  i was wondering about this, should have been more explicit.  anyway, giving the whole thing a look now.

File: src/busio.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ def __init__(self, scl, sda, frequency=400000):
 
     def init(self, scl, sda, frequency):
         self.deinit()
-        if board_id == "raspi_3" or board_id == "raspi_2" or board_id == "beaglebone_black":
+        if board_id == "raspi_3" or board_id == "raspi_2" or board_id == "beaglebone_black" or board_id == "orangepipc":
```
Comment (brennen) [inline src/busio.py]: might be more pythonic here to say `if board_id in ("raspi_3", "raspi_2", "beaglebone_black", "orangepipc")`.

File: src/busio.py [hunk 2]
```diff
@@ -69,23 +69,16 @@ def writeto_then_readfrom(self, address, buffer_out, buffer_in, *, out_start=0,
 
 class SPI(Lockable):
     def __init__(self, clock, MOSI=None, MISO=None):
-        print("SPI(): __init()")
         self.deinit()
-        if board_id == "raspi_3" or board_id == "raspi_2" or board_id == "beaglebone_black":
+        if board_id == "raspi_3" or board_id == "raspi_2" or board_id == "beaglebone_black" or board_id == "orangepipc":
```
Comment (brennen) [inline src/busio.py]: same

Comment (ladyada) [reply src/busio.py]: ditto!

**Comment by ladyada** [comment]
@dhalbert yah that code is about to be completely redone anyhow, this is just a placeholder till the platform detection helper is written this week.

**Comment by ladyada** [comment]
ok tested branch on a raspberry pi with my Ultimate HAT Of All Peripherals and they all still work

**Comment by Tonymac32** [comment]
I can try this out on some other Allwinner boards as well (I don't have this specific board), this looks simple enough as far as adding support is concerned, I think you'll find some more interest for FriendlElec NanoPi Duo  and NEO family, they are much more embedded, microcontroller-style hardware (headless/etc).

**Comment by ladyada** [comment]
@Tonymac32 that would be amazing. its very easy now that we have the allwinner chip defined  - if you can wait a couple hours, we'll get this merged and then you can submit PR's for boards you've got?

**Comment by ladyada** [comment]
yes! thank you everyone :) good news is with `libgpiod`, `pyserial`, `spidev`, and `SMbus` - there should be no low level code required. ill be adding 'soft PWM' out and 'soft pulsein' (already done mostly) using libgpiod C helpers as well, for chips that dont have PWM out.

**Comment by Tonymac32** [comment]
@ladyada It would be a few hours in any case, I am at "the day job", my Linux hackery hasn't gotten past "self-sustaining hobby" status so far...  I maintain the Amlogic boards and the Asus Tinker board for Armbian as well, I'll see what I can do, we're trying to get 4.19 LTS kernels rolled out/debugged.

**Comment by ladyada** [comment]
@Tonymac32 awesome - thanks for anything you can PR :)

**Comment by ladyada** [comment]
bombs away! will do a release, i did do testing on both orangepi and raspi but we'll see if i made some mistake - hopefully not :/

**Comment by jerryneedell** [comment]
Updated to 4.0.0 on my RPI 3B+  Ran a few things using SPI and I2C. No problems.

**Comment by ladyada** [comment]
thank ya!

**Comment by pdp7** [comment]
@ladyada @brennen would you be able to describe some use-cases for pull-up/pull-down flags on GPIO lines through libgpiod?  And would those ever be used without the open-source or open-drain flags?  Maybe a push button with no external resistor?

I just spoke with [Linus Walleij at the Linux Plumbers Conference](https://www.spinics.net/lists/linux-gpio/msg42377.html).  He is the maintainer of the gpio and pinctrl subsystems in the Linux kernel.  Linus thought it may be a good idea to expose pull-up/pull-down flags to the gpiolib userspace ABI so that libgpiod can use it.

I talked to Linus about how I could about creating a Linux kernel patch to achieve this and I'll speak further with [Bartosz Golaszewski, maintainer of libgpiod,](https://www.spinics.net/lists/linux-gpio/msg42448.html) at another conference next week.

In short:
Both Linus W. and Bartosz are interested in better understanding the use-case for pull-up/pull-down, so I'm hoping you might be able to help me explain it to them.

thanks!
_let me know if I should move this discussion to a new issue_

**Comment by ladyada** [comment]
hiya yes there was an issue at https://github.com/brgl/libgpiod/issues/31 which described use case - it seems that it was deleted 
the primary use is to avoid floating inputs from sensors and devices that do not always force the output - for example buttons without pullup/down resistors, or photo transistors like these breakbeam sensors. https://www.adafruit.com/product/2168
they can also be used for making bi-directional data transfer systems (i2c is an example) - you can mimic an open-drain input with an input and pullup resistor.

see also https://github.com/dotnet/iot/issues/410

**Comment by pdp7** [comment]
@ladyada Thank you for the insights.  The discussion of these change has continued on the linux kernel gpio subsystem mailing list.  The kernel gpio maintainer, Linus Walleij, and libgpiod author, @brgl, and @warthog618 have been discussing whether or not a pull-up/pull-down flag is sufficient as some chips support setting the resistance value for pull-up/pull-down:
https://lore.kernel.org/linux-gpio/CACRpkdZjswY4zW232ahSQSGfprbgBx8YL4Wb0i3ebegT00v3jQ@mail.gmail.com/

Do you think it is sufficient for libgpiod to just have pull-up/down flags that would use the chip's default resistance setting?

**Comment by ladyada** [comment]
huh tough question. while i have definitely seen high/med/low strength output pins, i dont recall ever seeing a chip with variable pullup/down strength. i think you can get away with having a generic ioctl for now, can always add more specialized ones if chip fabs change!
